lib/execution_engine2/db/MongoUtil.py

```diff
@@ -420,6 +447,26 @@ def update_job_status(self, job_id, status, msg=None, error_message=None):
     def mongo_engine_connection(self):
         yield self.me_connection
 
+    def insert_jobs(self, jobs_to_insert: List[Job]) -> List[ObjectId]:
+        """
+        Insert multiple job records using MongoEngine
+        :param jobs_to_insert: Multiple jobs to insert at once
+        :return: List of job ids from the insertion
+        """
+        # TODO Look at pymongo write_concerns that may be useful
+        # TODO see if pymongo is faster
+
+        inserted = Job.objects.insert(doc_or_docs=jobs_to_insert, load_bulk=False)
```

Comment:
Can you remind me how insert failures work? Can you get partial successes? How would partial failures be handled in the calling method?

Reply:
Before mongo 4.2 there aren't transactions, so yes there are partial successes and failures. If there is a failure, there should be some records inserted and some nothing, and then the method returns. A "created" jobs reaper could reap jobs that are in created state for too long.

Reply:
Say I submit 10 jobs and one fails. How does the code handle that, and what would the client programmer see?

Reply:
I'd expect a mongoengine error text to be shown in the ServerError

Reply:
Ok, so basically the 9 created jobs are abandoned and some sort of reaper would have to terminate them later I guess?

Reply:
Right,  this is covered by the reaper,  unless we can figure out how to terminate the jobs right here and then, or rollback them somehow.

Reply:
Well, the code could check if the inserted count equals the incoming job count and 

* retry any failed jobs for some number of retries
* remove all the non-failed jobs from the db and throw an error
* terminate all the non-failed jobs

Since this is job creation we're still in synchronous mode so I think all of those are safe options at least from the standpoint of feedback to the user. Of course if the DB is down or something and stays down there's not much that can be done

Reply:
Seems like a lot of work for this PR, maybe this should go into another PR?

Reply:
I'm not actually advocating for any particular solution right now. Let me ask this: what would happen in a partial failure situation assuming we were running this code?

Reply:
It calls collection.insert_many
```
        with set_write_concern(self._collection, write_concern) as collection:
            insert_func = collection.insert_many
            if return_one:
                raw = raw[0]
                insert_func = collection.insert_one

        try:
            inserted_result = insert_func(raw)
            ids = [inserted_result.inserted_id] if return_one else inserted_result.inserted_ids
        except pymongo.errors.DuplicateKeyError as err:
            message = 'Could not save document (%s)'
            raise NotUniqueError(message % six.text_type(err))
        except pymongo.errors.BulkWriteError as err:
            # inserting documents that already have an _id field will
            # give huge performance debt or raise
            message = 'Document must not have _id value before bulk write (%s)'
            raise NotUniqueError(message % six.text_type(err))
        except pymongo.errors.OperationFailure as err:
            message = 'Could not save document (%s)'
            if re.match('^E1100[01] duplicate key', six.text_type(err)):
                # E11000 - duplicate key error index
                # E11001 - duplicate key on update
                message = 'Tried to save duplicate unique keys (%s)'
                raise NotUniqueError(message % six.text_type(err))
            raise OperationError(message % six.text_type(err))
```
And fails, which can leave the db in a state with some jobs inserted, and some not inserted.  As for what happens when insert_many is called is

```
        blk = _Bulk(self, ordered, bypass_document_validation)
        blk.ops = [doc for doc in gen()]
        blk.execute(write_concern, session=session)
        return InsertManyResult(inserted_ids, write_concern.acknowledged)
```

Reply:
I mean what would happen to the jobs, and what would happen for the user

Reply:
The user gets a failure message during the submit, and some of the jobs are saved partially in a created state, and stay in that state until a reaper gets to them.

Reply:
Does the reaper exist now, or is that a future PR? How often does the reaper run?

Reply:
It [exists](https://github.com/kbase/execution_engine2/pull/412/files), just is not deployed and set up to run every X mins.

Reply:
Ok, so you're working on it

Reply:
All right, do you think leaving things as is is reasonable assuming the reaper gets done?

Reply:
yes

Reply:
:+1: